File: Adaline/C++/Adaline.cpp
#include "headers/Adaline.h"
using namespace std;

Adaline::Adaline(int input_size, double lr) : neuron(input_size), lr(lr) {}

Adaline::~Adaline() {}

void Adaline::learning(const std::vector<double> &x, const int y) {

  vector<double> w = this->neuron.getW();
  double b = this->neuron.getB();

  double u = this->neuron.soma(x);

  for (int i = 0; i < x.size(); ++i) {
    w[i] = w[i] + this->lr * (y - u) * x[i];
  }
  b = b + this->lr * (y - u);

  this->neuron.setW(w);
  this->neuron.setB(b);
}

int Adaline::predict(const std::vector<double> &x) {

  double u = this->neuron.soma(x);

  return this->neuron.activation(u);
}

void Adaline::train(const std::vector<std::vector<double>> &X,
                    const std::vector<int> &y, const int num_epochs) {
  for (int epoch = 0; epoch < num_epochs; epoch++) {
    for (int i = 0; i < X.size(); i++) {
      this->learning(X[i], y[i]);
    }
  }
}

File: Adaline/Python/src/adaline.py
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
from src.neuron import Neuron 

class Adaline:
    def __init__(self, num_inputs:int, lr:float):
        self.neuron = Neuron(num_inputs)
        self.lr = lr 

    def learning(self, lr: float, y: float, x: np.ndarray): # Delta rule
        w = self.neuron.getW()
        b = self.neurongetB()
        u = self.neuron.soma(x)

        w = w + lr * (y - u) * x
        b = b + lr * (y - u)

        self.neuron.setW(w)
        self.neuron.setB(b)

    def predict(self, x:np.ndarray):
        u = self.neuron.soma(x)
        return self.neuron.activation(u)

    def train(self, X: np.ndarray, y: np.ndarray, num_epochs: int):
        for epoch in range(num_epochs):
            for i in range(len(X)):
                self.learning(self.lr, y[i], X[i])
